pkg/patterns/addon/pkg/loaders/git.go

```diff
@@ -0,0 +1,130 @@
+package loaders
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-git/go-git/v5"
+	"io/ioutil"
+	"os"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+type GitRepository struct {
+	baseURL string
+	subDir string
+	branch string
+}
+
+var _ Repository = &GitRepository{}
+
+// NewGitRepository constructs an GitRepository
+func NewGitRepository(baseurl string) *GitRepository{
+	repo := parseGitURL(baseurl)
+	return &repo
+}
+
+func (r *GitRepository) LoadChannel(ctx context.Context,  name string) (*Channel, error){
+	if !allowedChannelName(name) {
+		return nil, fmt.Errorf("invalid channel name: %q", name)
+	}
+
+	log := log.Log
+	log.WithValues("baseURL", r.baseURL).Info("loading channel")
+	log.WithValues("baseURL", r.baseURL).Info("cloning git repository")
+
+	if r.subDir != "" {
+		name = r.subDir + "/" + name
+	}
+	b, err := r.readURL(name)
+	if err != nil{
+		log.WithValues("path", name).Error(err, "error reading channel")
+		return nil, err
+	}
+	fmt.Println(string(b))
+
+	channel := &Channel{}
+	if err := yaml.Unmarshal(b, channel); err != nil {
+		return nil, fmt.Errorf("error parsing channel %s: %v", name, err)
+	}
+
+	return channel, nil
+}
+
+
+func (r *GitRepository) LoadManifest(ctx context.Context, packageName string, id string) (map[string]string, error){
+	if !allowedManifestId(packageName) {
+		return nil, fmt.Errorf("invalid package name: %q", id)
+	}
+
+	if !allowedManifestId(id) {
+		return nil, fmt.Errorf("invalid manifest id: %q", id)
+	}
+
+	log := log.Log
+	log.WithValues("package", packageName).Info("loading package")
+
+	var filePath string
+	if r.subDir == "" {
+		filePath = fmt.Sprintf("packages/%v/%v/manifest.yaml", packageName, id)
```

Comment:
If you use `path.Join` you can probably avoid the repetition here: https://play.golang.org/p/-XheAesp5eZ

(Note: path.Join and not filepath.Join, because even on windows we want a url-style path with forward slashes)